File: models/luban.go
package models

import (
	"time"
)

// ConfigProject 配置项目表
type ConfigProject struct {
	ID          uint      `json:"id" gorm:"primaryKey"`
	Name        string    `json:"name" gorm:"size:100;not null"`          // 项目名称
	Description string    `json:"description" gorm:"type:text"`           // 项目描述
	RootPath    string    `json:"root_path" gorm:"size:255;not null"`     // 配置根目录
	OutputPath  string    `json:"output_path" gorm:"size:255;not null"`   // 输出目录
	Status      string    `json:"status" gorm:"size:20;default:'active'"` // 状态：active, archived
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// ConfigTable 配置表定义
type ConfigTable struct {
	ID          uint      `json:"id" gorm:"primaryKey"`
	ProjectID   uint      `json:"project_id" gorm:"index"`                // 所属项目ID
	Name        string    `json:"name" gorm:"size:100;not null"`          // 配置表名称
	Description string    `json:"description" gorm:"type:text"`           // 配置表描述
	FilePath    string    `json:"file_path" gorm:"size:255;not null"`     // 配置文件路径
	FileType    string    `json:"file_type" gorm:"size:20;not null"`      // 文件类型：excel, csv, json, xml, yaml
	SheetName   string    `json:"sheet_name" gorm:"size:100"`             // Excel工作表名称
	Validators  string    `json:"validators" gorm:"type:text"`            // 数据验证规则（JSON格式）
	Status      string    `json:"status" gorm:"size:20;default:'active'"` // 状态：active, archived
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// ConfigExport 配置导出记录
type ConfigExport struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	ProjectID uint      `json:"project_id" gorm:"index"`                 // 所属项目ID
	TableID   uint      `json:"table_id" gorm:"index"`                   // 配置表ID
	Format    string    `json:"format" gorm:"size:20;not null"`          // 导出格式：binary, json, bson, xml, lua, yaml
	Language  string    `json:"language" gorm:"size:20;not null"`        // 目标语言：cpp, java, go, cs, python, lua
	Status    string    `json:"status" gorm:"size:20;default:'pending'"` // 状态：pending, running, success, failed
	Output    string    `json:"output" gorm:"type:text"`                 // 导出输出
	Error     string    `json:"error" gorm:"type:text"`                  // 错误信息
	StartTime time.Time `json:"start_time"`                              // 开始时间
	EndTime   time.Time `json:"end_time"`                                // 结束时间
	Duration  int       `json:"duration"`                                // 执行时长(秒)
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

File: public/js/admin/luban.js
function lubanManagement() {
    return {
        projects: [],
        currentProject: null,
        currentProjectTables: [],
        showProjectModal: false,
        showProjectDetailModal: false,
        showTableModal: false,
        showExportModal: false,
        editMode: false,
        form: {
            id: '',
            name: '',
            description: '',
            root_path: '',
            output_path: '',
            status: 'active'
        },
        tableForm: {
            id: '',
            project_id: '',
            name: '',
            description: '',
            file_path: '',
            file_type: 'excel',
            sheet_name: '',
            validators: '',
            status: 'active'
        },
        exportForm: {
            project_id: '',
            table_id: '',
            format: 'json',
            language: 'go'
        },
        exportProgress: null,
        progressTimer: null,

        init() {
            this.fetchProjects();
        },

        async fetchProjects() {
            try {
                const response = await fetch('/api/admin/luban/projects');
                if (!response.ok) throw new Error('获取项目列表失败');
                this.projects = await response.json();
            } catch (error) {
                ShowError(error.message);
            }
        },

        createProject() {
            this.editMode = false;
            this.form = {
                id: '',
                name: '',
                description: '',
                root_path: '',
                output_path: '',
                status: 'active'
            };
            this.showProjectModal = true;
        },

        editProject(project) {
            this.editMode = true;
            this.form = { ...project };
            this.showProjectModal = true;
        },

        async submitProject() {
            try {
                const url = this.editMode ? `/api/admin/luban/projects/${this.form.id}` : '/api/admin/luban/projects';
                const method = this.editMode ? 'PUT' : 'POST';
                
                const formData = { ...this.form };
                if (!this.editMode) {
                    delete formData.id;
                }
                
                const response = await fetch(url, {
                    method,
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify(formData)
                });

                if (!response.ok) {
                    const error = await response.json();
                    throw new Error(error.error || '操作失败');
                }

                await this.fetchProjects();
                this.showProjectModal = false;
                ShowMessage(this.editMode ? '项目更新成功' : '项目创建成功');
            } catch (error) {
                ShowError(error.message);
            }
        },

        async deleteProject(id) {
            if (!confirm('确定要删除这个项目吗？')) return;

            try {
                const response = await fetch(`/api/admin/luban/projects/${id}`, {
                    method: 'DELETE',
                });

                if (!response.ok) throw new Error('删除项目失败');

                await this.fetchProjects();
                ShowMessage('项目删除成功');
            } catch (error) {
                ShowError(error.message);
            }
        },

        async viewProject(project) {
            this.currentProject = project;
            await this.fetchProjectTables(project.id);
            this.showProjectDetailModal = true;
        },

        async fetchProjectTables(projectId) {
            try {
                const response = await fetch(`/api/admin/luban/tables?project_id=${projectId}`);
                if (!response.ok) throw new Error('获取配置表列表失败');
                this.currentProjectTables = await response.json();
            } catch (error) {
                ShowError(error.message);
            }
        },

        createTable() {
            this.editMode = false;
            this.tableForm = {
                id: '',
                project_id: this.currentProject.id,
                name: '',
                description: '',
                file_path: '',
                file_type: 'excel',
                sheet_name: '',
                validators: '',
                status: 'active'
            };
            this.showTableModal = true;
        },

        editTable(table) {
            this.editMode = true;
            this.tableForm = { ...table };
            this.showTableModal = true;
        },

        async submitTable() {
            try {
                const url = this.editMode 
                    ? `/api/admin/luban/tables/${this.tableForm.id}` 
                    : '/api/admin/luban/tables';
                const method = this.editMode ? 'PUT' : 'POST';
                
                const formData = { ...this.tableForm };
                if (!this.editMode) {
                    delete formData.id;
                }
                
                const response = await fetch(url, {
                    method,
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify(formData)
                });

                if (!response.ok) {
                    const error = await response.json();
                    throw new Error(error.error || '操作失败');
                }

                await this.fetchProjectTables(this.currentProject.id);
                this.showTableModal = false;
                ShowMessage(this.editMode ? '配置表更新成功' : '配置表创建成功');
            } catch (error) {
                ShowError(error.message);
            }
        },

        async deleteTable(id) {
            if (!confirm('确定要删除这个配置表吗？')) return;

            try {
                const response = await fetch(`/api/admin/luban/tables/${id}`, {
                    method: 'DELETE',
                });

                if (!response.ok) throw new Error('删除配置表失败');

                await this.fetchProjectTables(this.currentProject.id);
                ShowMessage('配置表删除成功');
            } catch (error) {
                ShowError(error.message);
            }
        },

        exportTable(table) {
            this.exportForm = {
                project_id: this.currentProject.id,
                table_id: table.id,
                format: 'json',
                language: 'go'
            };
            this.exportProgress = null;
            this.showExportModal = true;
        },

        async submitExport() {
            try {
                const response = await fetch('/api/admin/luban/export', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify(this.exportForm)
                });

                if (!response.ok) {
                    const error = await response.json();
                    throw new Error(error.error || '操作失败');
                }

                const result = await response.json();
                this.startProgressPolling(result.id);
                ShowMessage('导出任务已开始');
            } catch (error) {
                ShowError(error.message);
            }
        },

        startProgressPolling(exportId) {
            if (this.progressTimer) {
                clearInterval(this.progressTimer);
            }

            const pollProgress = async () => {
                try {
                    const response = await fetch(`/api/admin/luban/exports/progress/${exportId}`);
                    if (!response.ok) throw new Error('获取导出进度失败');
                    
                    this.exportProgress = await response.json();
                    
                    if (this.exportProgress.status !== 'running') {
                        clearInterval(this.progressTimer);
                        this.progressTimer = null;

                        if (this.exportProgress.status === 'success') {
                            ShowMessage('导出成功');
                        } else if (this.exportProgress.status === 'failed') {
                            ShowError('导出失败: ' + this.exportProgress.error);
                        }
                    }
                } catch (error) {
                    clearInterval(this.progressTimer);
                    this.progressTimer = null;
                    ShowError(error.message);
                }
            };

            this.progressTimer = setInterval(pollProgress, 1000);
            pollProgress(); // 立即执行一次
        }
    };
} 